server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java

```diff
@@ -585,71 +604,69 @@ static long fallbackRegionSize(JvmInfo jvmInfo) {
-                        long localBlackHole = 0;
-                        // number of allocations, corresponding to (approximately) number of free regions + 1
-                        int allocationCount = Math.toIntExact((maxHeap - memoryUsed.baseUsage) / g1RegionSize + 1);
-                        // allocations of half-region size becomes single humongous alloc, thus taking up a full region.
-                        int allocationSize = (int) (g1RegionSize >> 1);
-                        long maxUsageObserved = memoryUsed.baseUsage;
-                        for (; allocationIndex < allocationCount; ++allocationIndex) {
-                            long current = currentMemoryUsageSupplier.getAsLong();
-                            if (current >= maxUsageObserved) {
-                                maxUsageObserved = current;
-                            } else {
-                                // we observed a memory drop, so some GC must have occurred
-                                break;
-                            }
-                            if (initialCollectionCount != gcCountSupplier.getAsLong()) {
-                                break;
-                            }
-                            localBlackHole += new byte[allocationSize].hashCode();
-                        }
-
-                        blackHole += localBlackHole;
-                        logger.trace("black hole [{}]", blackHole);
-
-                        long now = timeSupplier.getAsLong();
-                        this.lastCheckTime = now;
-                        allocationDuration = now - begin;
-                        this.attemptNo = 0;
-                    }
+                    result = tryTriggerGC(memoryUsed);
+                } else {
+                    logger.info("could not acquire lock within {} when attempting to trigger G1GC due to high heap usage", lockTimeout);
                 }
             } catch (InterruptedException e) {
                 logger.info("could not acquire lock when attempting to trigger G1GC due to high heap usage");
                 Thread.currentThread().interrupt();
                 // fallthrough
             }
 
+            if (performingFullGC && attemptNoCopy == 0) {
+                // Another thread is currently performing a full GC, and we were not able to try (lock acquire timeout)
+                // Since the full GC thread may hold the lock for longer, try again with an increased timeout
+                logger.info(
+                    "could not acquire lock within {} while another thread was performing a full GC, waiting again for {}",
+                    lockTimeout,
+                    fullGCLockTimeout
+                );
+                try (ReleasableLock locked = lock.tryAcquire(fullGCLockTimeout)) {
+                    if (locked != null) {
+                        attemptNoCopy = ++this.attemptNo;
```

Comment:
Should we add a condition here, where we try to trigger a GC only if the situation did not get better? It could make sense: we know we timed out for a full GC; if meanwhile the situation got better we can skip things. 
May be redundant, we might return fast due to the lastCheckTime + minimumInterval, but a full GC may be > minimumInterval

Reply:
Resolved by nesting full GC condition inside "normal" triggering: memory usage is examined before attempting pretty much anything and lead to an early exit